packages/kit/src/utils/routing.js

```diff
@@ -143,3 +149,20 @@ export function exec(match, { names, types, optional }, matchers) {
 
 	return params;
 }
+
+/** @param {string} str */
+function escape(str) {
+	return (
+		str
+			.normalize()
+			// escape [ and ] before escaping other characters, since they are used in the replacements
+			.replace(/[[\]]/g, '\\$&')
+			// replace %, /, ? and # with their encoded versions
```

Comment:
```suggestion
			// replace %, /, ? and # with their encoded versions because decode_pathname leaves them untouched
```

Reply:
also, would you mind updating the JSDoc for `decode_pathname`. it says something like "to prevent double decoding", but should really just say "to prevent decoding"

Reply:
hang on, do we need `decode_pathname`? can't we just do `decodeURI`?

Reply:
oh, yeah, I think you're probably right :laughing:

Reply:
probably it didn't work because of Vite

Reply:
it _does_ prevent double decoding though — it stops `/%2523` from turning into `/#` instead of `%23`, which is what would happen if we did `decodeURI` followed by `decode_params` instead of `decode_pathname` followed by `decode_params`